main.cpp

```diff
@@ -1,75 +1,96 @@
-};
-
-class MyASTConsumer : public ASTConsumer {
-public:
-    MyASTConsumer(Rewriter &rewriter) : callback_(rewriter) {
-        matcher_.addMatcher(
-                cStyleCastExpr(unless(isExpansionInSystemHeader())).bind("cast"), &callback_);
-    }
-
-    void HandleTranslationUnit(ASTContext &Context) override {
-        matcher_.matchAST(Context);
-    }
-
-private:
-    CastCallBack callback_;
-    MatchFinder matcher_;
-};
-
-class CStyleCheckerFrontendAction : public ASTFrontendAction {
-public:
-    CStyleCheckerFrontendAction() = default;
-    
-    void EndSourceFileAction() override {
-        rewriter_.getEditBuffer(rewriter_.getSourceMgr().getMainFileID())
-            .write(llvm::outs());
-    }
-
-    std::unique_ptr<ASTConsumer> CreateASTConsumer(CompilerInstance &CI, StringRef /* file */) override {
-        rewriter_.setSourceMgr(CI.getSourceManager(), CI.getLangOpts());
-        return std::make_unique<MyASTConsumer>(rewriter_);
-    }
-
-private:
-    Rewriter rewriter_;
-};
-
-static llvm::cl::OptionCategory CastMatcherCategory("cast-matcher options");
-
-int main(int argc, const char **argv) {
-    // Ubuntu 20
-    auto Parser = llvm::ExitOnError()(CommonOptionsParser::create(argc, argv, CastMatcherCategory));
-    // For Ubuntu 18, use the old Clang API to construct the CommonOptionsParser:
-    // CommonOptionsParser Parser(argc, argv, CastMatcherCategory);
-
-    ClangTool Tool(Parser.getCompilations(), Parser.getSourcePathList());
-    return Tool.run(newFrontendActionFactory<CStyleCheckerFrontendAction>().get());
-}
+#include <iostream>
+#include <sstream>
+
+#include <llvm/Support/CommandLine.h>
+#include <llvm/Support/Error.h>
+
+#include <clang/ASTMatchers/ASTMatchFinder.h>
+#include <clang/ASTMatchers/ASTMatchers.h>
+#include <clang/Basic/Diagnostic.h>
+#include <clang/Frontend/CompilerInstance.h>
+#include <clang/Frontend/FrontendActions.h>
+#include <clang/Rewrite/Core/Rewriter.h>
+#include <clang/Tooling/CommonOptionsParser.h>
+#include <clang/Tooling/Tooling.h>
+
+using namespace clang;
+using namespace clang::ast_matchers;
+using namespace clang::tooling;
+
+class CastCallBack : public MatchFinder::MatchCallback {
+
+private:
+  Rewriter& _rewriter;
+
+public:
+  CastCallBack(Rewriter &rewriter) : _rewriter(rewriter){};
+  void run(const MatchFinder::MatchResult &Result) override {
+
+    const auto *castExpr = Result.Nodes.getNodeAs<CStyleCastExpr>("cast");
+    
+    const auto range = CharSourceRange::getCharRange(
+        castExpr->getLParenLoc(),
+        castExpr->getSubExprAsWritten()->getBeginLoc());
+
+    const auto &sourceManager = *Result.SourceManager;
+
+    const auto type = Lexer::getSourceText(
+        CharSourceRange::getTokenRange(
+            castExpr->getLParenLoc().getLocWithOffset(1),
+            castExpr->getRParenLoc().getLocWithOffset(-1)),
+                             sourceManager, LangOptions());
+
+    const auto *expr = castExpr->getSubExprAsWritten()->IgnoreImpCasts();
```

Comment:
Please describe why you are using here `getSubExprAsWritten()` and `IgnoreImpCasts()`.
What would be the difference with the use of `getSubExprAsWritten()->IgnoreImpCasts()`, `getSubExpr()->IgnoreImpCasts()` and `getSubExprAsWritten()`?

Reply:
getSubExprAsWritten() gets us the cast subexpression as it was written in the source code, looking through any implicit casts or other intermediate nodes introduced by semantic analysis. Then we use IgnoreImpCasts() to ignore any implicit casts which might surround this expression.

Talking about our example in particular, it wouldn't matter if we used `getSubExprAsWritten()` alone or with IgnoreImpCasts(). getSubExprAsWritten()->IgnoreImpCasts(), getSubExpr()->IgnoreImpCasts() and getSubExprAsWritten() would all give us the same result.